### DISCUSSION TITLE
MONAI Label, how to start a remote server and connect to it via a remote ip.

### DISCUSSION THREAD

**Discussion by lenlobo** [Project-MONAI/MONAILabel#1379] [Q&A] [answered]
I have a ML Workspace or a server that I work on, I installed MONAI Label on it. 
Now I have two scenarios,
First - I want to start the MONAI Label, by running the monailabel ` start_server ...` command, but want to use 3d slicer on my local machine.
Second - I want to start the MONAI Label, on the remote server, and then run a docker container of slicer and connect to the monai label server, completely remotely.

Is this even possible? If yes, some insights would be appreciated.

Futher more, is using a pretrained model and then labeling the new data possible by using 3dslicer with monailabel?
The scenario here is, I have a model which is trained on some data, I kind of want to use the model to annotate or label the new data.

**Comment by tangy5** [comment]
Hi @lenlobo , yes, both scenarios should work, I typically use the same way, but you may need to double-check several things: 

For the first scenario: yes, but please make sure the remote and local machines are in the same network. This is typically achieved by connecting to VPN or other networking revolvers. And make sure your local machine can reach remote machine host/IP

Second, this should be working, MONAI Label won't bother Slicer's docker, unless you can start Slicer, then you can type the host and port of MONAI Label server, then use it in Slicer. Same as the first scenario, it's needed to resolve networking first, in the same network. 

For the additional question, if new data are the same target classes (no new segments), there is no problem you can fine-tune the pretrained mode. If you are adding new classes/organs/tissues, you need to change and define a network, then train it. 

Enjoy MONAI Label

**Reply by lenlobo** [reply]
Hey @tangy5

I have use my organisation's server, I installed MONAI Label using a remote terminal, so that has been setup there, I am confused with how (or which) to start the docker contrainer of 3d slicer, install the monailabel extension, and then connect. 
I started a docker contrainer with slicer running on jupyter and then ran a code snippet, which starts an interactive UI of Slicer on the browser, but this doesn't allow me to install any extensions, so hence I am not able to run MONAI. If you have tried this before, if possible a step by step guide would be highly appreciated, or atleast a link to an article as well!

For the MONAI Server on remote and 3d on Local, how do I start the server on a specific IP? I am not sure that the Server which will run remotely MONAI Label, and the me working from my home, connecting to an VPN, should be on the same network. Anyway, but its still the question that, how will I start the server on a specific IP, or lets say the server has an IP which I know, can I assign a port while I start it, so that I can just connect it locally?

Thanks!

**Reply by tangy5** [reply]
Hi @lenlobo , I don't have much experience on using Slicer in Docker, probably the Slicer community can answer this question better. 

For local Slicer, after you installed the MONAI label plugin, you could see there is a box named "MONAI Label server", you can type host and IP there.

**Reply by lenlobo** [reply]
So I'll tell what I have done,
1. Logged into my server using an SSH Terminal, did `monailabel start_server --host x.y.z --port xxxx -app apps/radiology --studies path/to/data/ --conf models segmentation`
2. This runs and starts the server at xyz:xxxx. Now xyz is the same server I ssh and connect to earlier, its my remote server my organisation uses.
3. Now, I go to Slicer > Monai Label extenstion on my Local System. I put in the ip, xyz:xxxx, and it says check the server uri. and stops working.

I can ping the xyz server in my terminal(local), but I cannot ping xyz:xxxx. Are you also using this setup for your work or is it a server you've setup? I am not sure if this is due to security or the server being setup incorrectly in my side.

**Reply by tangy5** [reply]
I can't tell the issue here, but you can try this:
Directly run this without host and IP
`monailabel start_server -app apps/radiology --studies path/to/data/ --conf models segmentation`
MONAI Label server will run at http://0.0.0.0:8000 at the remote machine. 

Then you can check the remote machine's IP. 
In your local Slicer, type "http://<remote_IP>:8000", replace the real remote machine IP

This works for me when I need to access remote MONAI Label server.

**Reply by lenlobo** [reply]
This is what I end up with doing this, I tried without the port, but that was under use, so tried --port 8080.
![image](https://user-images.githubusercontent.com/46513163/230672131-05cdb99a-0b95-462d-8a85-0829b9ca77a3.png)

**Reply by tangy5** [reply]
If you can confirm the server_url is correct, then this probably means you can't reach the server uri...
You might need to check and consult the networking issues. 
You can try this if you'd like try this remote setting for MONAI Label, I have a server up on my side using a public accessible machine for GTC demo purpose. http://44.204.23.92:8000   | This is only available for 2 hours.

**Reply by lenlobo** [reply]
Okay, thanks for the insights! I would look into it :), Happy Weekend and Happy Easter in advance if you celebrate it

**Reply by tangy5** [reply]
Thanks. If you want to try this, see the above message.

**Reply by lenlobo** [reply]
So I can connect it using my personal laptop, but cannot on my work system. But thats not the case with my work server, I can't connect to it on both :/.

**Reply by tangy5** [reply]
Yup, hope this can help you solve the networking issue. Good luck.

**Reply by diazandr3s** [reply]
It could be a firewall issue. You may have some ports blocked. In order to access to remote MONAI Label server, you have open the port MONAI Label is using (i.e 8000)

**Reply by lenlobo** [reply]
@diazandr3s Hi,

Is it mandatory to use port 8000? Because I've tried different ports by `--port xxxx` which are not busy, and they didnt seem to work. Because I can ping the domain name, or the ip address of the GPU remote server, but cannot ping it with the port. 
I am asking is it mandatory to use 8000 because it seems to be busy at the moment on the server.

**Reply by diazandr3s** [reply]
No, you could start the MONAI Label server on any port. Just specify it with `--port` argument

**Reply by lenlobo** [reply]
Yeah, I've tried it but no luck, it doesn't atill connect, probably some firewall issues, can you point me somewhere I can figure what can the issue be? Or I can raise an issue with my team about something?

**Comment by SachidanandAlle** [comment]
`monailabel start_server -h`
and yes.. u don't need to run on 8000 port.. if u have other port opened up in your firewall and accessible.. u can use that too..
for example you can use 8080 instead..

and when you start the server.. it prints a log.. you can check the log.. something like this..
`Uvicorn running on http://0.0.0.0:8090`

**Reply by lenlobo** [reply]
Yeah, not sure why but I can't seem to connect using the ip:port whatever I mention, the remote server is a GPU server my organization uses.

**Comment by SachidanandAlle** [comment]
if you are running docker.. you want to make sure you have port mapping correctly done.. 
forget about monailabel server.. you can simple debug what's wrong with your vpn/remote/docker setup..

for example.. try to run simple http server in python
`python3 -m http.server 8090`

Once you are able to access this.. you can go back to monailabel and run it..

**Reply by lenlobo** [reply]
Okay, will give this debug a try!

**Reply by lenlobo** [reply]
Hi @SachidanandAlle ,
I tried what you said, I don't really know am I supposed to get an output from it? because It was just stuck `Serving HTTP on :: port 8080 (http://[::]:8080/) ..`

I just wanted to say, I started a docker container on the same GPU Server with a port 8080 and I can access that container by going on to that IP:8080, and it works, but it is an issue when I start the monai server. A question, to connect, should it be only an IP or a domain name is fine as well? I have tried both didn't work but still just asking.

**Reply by diazandr3s** [reply]
Hi @lenlobo,

Could you paste the command you're using to start the Docker container?  If you want to map the port, you should start the container like this:

`docker run --gpus all --rm -ti --ipc=host -p 8080:8000 --net=host projectmonai/monailabel:latest bash`

This means you map port 8000 in the container to port 8080 on the Docker host.

You should use a port that is open in the firewall.

Then you can start the MONAI Label using the Docker container as usual.

**Reply by lenlobo** [reply]
Hi @diazandr3s ,
So I'm not using a monailabel docker container, I just installed Monailabel using the GPU server terminal using pip. 

Should I try to start a container instead of this approach?

**Comment by lenlobo** [comment, marked as answer]
https://github.com/Project-MONAI/MONAILabel/discussions/776#discussioncomment-2722416 
Hi,
The reply from Zrrr1997 seems to work for me at the moment. He suggested creating a ssh tunnel from my localport (eg 9090) to the port on which the monai label server is started. Worked!
